ADALiOS/ADALiOS/ADKeychainItem.m

```diff
@@ -0,0 +1,366 @@
+{
+    for (ADKeychainToken* token in _accessTokens)
+    {
+        // Ignore anything in the keychain item that's not the correct class.
+        if (![token isKindOfClass:[ADKeychainToken class]])
+        {
+            continue;
+        }
+        
+        if ([scopes isSubsetOfSet:token.scopes])
+        {
+            return token;
+        }
+    }
+    
+    return nil;
+}
+
+- (void)addAccessTokenWithScopes:(NSSet*)scopes
+                          toItem:(ADTokenCacheStoreItem*)item
+{
+    item.refreshToken = _refreshToken;
+    
+    ADKeychainToken* token = [self tokenWithScopes:scopes];
+    [token addToTokenItem:item];
+}
+
+- (void)removeIntersectingTokens:(NSSet*)scopes
+{
+    NSMutableIndexSet* toRemove = [NSMutableIndexSet indexSet];
+    
+    NSUInteger cTokens = [_accessTokens count];
+    
+    for (NSUInteger i = 0; i < cTokens; i++)
+    {
+        ADKeychainToken* token = _accessTokens[i];
+        
+        // Ignore anything in the keychain item that's not the correct class.
+        if (![token isKindOfClass:[ADKeychainToken class]])
+        {
+            continue;
+        }
+        
+        if ([scopes intersectsSet:token.scopes])
+        {
+            [toRemove addIndex:i];
+        }
+    }
+    
+    [_accessTokens removeObjectsAtIndexes:toRemove];
+}
+
+- (void)updateToTokenItem:(ADTokenCacheStoreItem*)item
+{
+    _refreshToken = item.refreshToken;
+    
+    NSSet* scopes = item.scopes;
+    [self removeIntersectingTokens:scopes];
+    
+    ADKeychainToken* token = [ADKeychainToken new];
+    [token updateToTokenItem:item];
+    [_accessTokens addObject:token];
+}
+
+@end
+
+@interface ADKeychainItem ()
+
+@property NSString* authority;
+@property NSString* clientId;
+@property NSData* sessionKey;
+@property ADProfileInfo* profileInfo;
+
+@end
+
+
+@implementation ADKeychainItem
+{
+    NSMutableDictionary* _policies;
+}
+
+
+
++ (ADKeychainItem*)itemForData:(NSData *)data
+{
+    if (!data)
+    {
+        return nil;
+    }
+    
+    id item = [NSKeyedUnarchiver unarchiveObjectWithData:data];
+    if (![item isKindOfClass:[ADKeychainItem class]])
+    {
+        return nil;
+    }
+    
+    return item;
+}
+
++ (BOOL)supportsSecureCoding
+{
+    return YES;
+}
+
+- (id)init
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    _policies = [NSMutableDictionary new];
+    
+    return self;
+}
+
+- (id)initWithCoder:(NSCoder*)aDecoder
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    _authority = [aDecoder decodeObjectOfClass:[NSString class] forKey:@"authority"];
+    _clientId = [aDecoder decodeObjectOfClass:[NSString class] forKey:@"clientId"];
+    _sessionKey = [aDecoder decodeObjectOfClass:[NSString class] forKey:@"sessionKey"];
+    _profileInfo = [aDecoder decodeObjectOfClass:[ADProfileInfo class] forKey:@"profileInfo"];
+    
+    _policies = [aDecoder decodeObjectOfClass:[NSArray class] forKey:@"policies"];
```

Comment:
This should be NSDictionary, not either array or set, as this is the keychain data object.